routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,381 @@
+#pragma once
+
+#include "routing/index_graph.hpp"
+#include "routing/joint.hpp"
+#include "routing/routing_exception.hpp"
+#include "routing/vehicle_mask.hpp"
+
+#include "coding/bit_streams.hpp"
+#include "coding/elias_coder.hpp"
+#include "coding/reader.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "std/cstdint.hpp"
+#include "std/limits.hpp"
+#include "std/unordered_map.hpp"
+#include "std/unordered_set.hpp"
+#include "std/vector.hpp"
+
+namespace routing
+{
+class IndexGraphSerializer final
+{
+public:
+  IndexGraphSerializer() = delete;
+
+  template <class Sink>
+  static void Serialize(IndexGraph const & graph,
+                        unordered_map<uint32_t, VehicleMask> const & masks, Sink & sink)
+  {
+    Header header(graph, masks);
+    JointIdEncoder jointEncoder;
+
+    vector<SectionSerializer> serializers;
+    PrepareSectionSerializers(graph, masks, serializers);
+
+    for (SectionSerializer & serializer: serializers)
+    {
+      Joint::Id const begin = jointEncoder.GetCount();
+      serializer.SerializeToBuffer(graph, masks, jointEncoder);
+      header.AddSection({serializer.GetMask(), serializer.GetBufferSize(), serializer.GetNumRoads(),
+                         begin, jointEncoder.GetCount()});
+    }
+
+    header.Serialize(sink);
+    for (SectionSerializer & section : serializers)
+      section.Flush(sink);
+  }
+
+  template <class Source>
+  static void Deserialize(IndexGraph & graph, Source & src, VehicleMask requiredMask)
+  {
+    Header header;
+    header.Deserialize(src);
+
+    JointIdConverter jointIdConverter(header.GetNumJoints());
+
+    for (uint32_t i = 0; i < header.GetNumSections(); ++i)
+    {
+      Section const & section = header.GetSection(i);
+      VehicleMask const mask = section.GetMask();
+
+      if (!(mask & requiredMask))
+      {
+        src.Skip(section.GetSize());
+        continue;
+      }
+
+      JointIdDecoder jointIdDecoder(section.GetBeginJointId());
+      BitReader<Source> reader(src);
+      uint64_t const expectedEndPos = src.Pos() + section.GetSize();
+
+      // -1 for uint32_t is some confusing, but it allows process first iteration in the common way.
+      // It works because uint32_t is modular ring type.
```

Comment:
Same as above - comment about why -1 allows you to encode zero-th feature id, and replace modular by residual.